In [1]:
import pandas as pd

In [2]:
bookings = pd.read_csv('bookings.csv', sep = ';')

In [3]:
bookings.head() 

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  

[5 rows x 21 columns]

In [4]:
bookings.shape

(119390, 21)

В файле 119390 строк и 21 столбец

In [5]:
bookings.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

Большинство переменных относится типу object и int64

In [6]:
# перевод названия столбцов к нижнему регистру и замена пробелов на знак нижнего подчёркивания

bookings.columns = bookings.columns.str.lower().str.replace(' ', '_')
bookings

hotel  is_canceled  lead_time arrival_full_date  \
0       Resort Hotel            0        342        2015-07-01   
1       Resort Hotel            0        737        2015-07-01   
2       Resort Hotel            0          7        2015-07-01   
3       Resort Hotel            0         13        2015-07-01   
4       Resort Hotel            0         14        2015-07-01   
...              ...          ...        ...               ...   
119385    City Hotel            0         23        2017-08-30   
119386    City Hotel            0        102        2017-08-31   
119387    City Hotel            0         34        2017-08-31   
119388    City Hotel            0        109        2017-08-31   
119389    City Hotel            0        205        2017-08-29   

        arrival_date_year arrival_date_month  arrival_date_week_number  \
0                    2015               July                        27   
1                    2015               July                        27   
2                    2015               July                        27   
3                    2015               July                        27   
4                    2015               July                        27   
...                   ...                ...                       ...   
119385               2017             August                        35   
119386               2017             August                        35   
119387               2017             August                        35   
119388               2017             August                        35   
119389               2017             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  ...  adults  children  babies  meal country  \
0                          0  ...       2       0.0       0    BB     PRT   
1                          0  ...       2       0.0       0    BB     PRT   
2                          1  ...       1       0.0       0    BB     GBR   
3                          1  ...       1       0.0       0    BB     GBR   
4                          2  ...       2       0.0       0    BB     GBR   
...                      ...  ...     ...       ...     ...   ...     ...   
119385                     5  ...       2       0.0       0    BB     BEL   
119386                     5  ...       3       0.0       0    BB     FRA   
119387                     5  ...       2       0.0       0    BB     DEU   
119388                     5  ...       2       0.0       0    BB     GBR   
119389                     7  ...       2       0.0       0    HB     DEU   

       reserved_room_type assigned_room_type customer_type reservation_status  \
0                       C                  C     Transient          Check-Out   
1                       C                  C     Transient          Check-Out   
2                       A                  C     Transient          Check-Out   
3                       A                  A     Transient          Check-Out   
4                       A                  A     Transient          Check-Out   
...                   ...                ...           ...                ...   
119385                  A                  A     Transient          Check-Out   
119386         

Пользователи из каких стран совершили наибольшее число успешных бронирований.
Бронирование считается успешным, если в дальнейшем оно не было отменено (переменная is_canceled).
Отобрать страны, входящие в топ-5.

In [7]:
max_country_bookings = bookings.query("is_canceled == 0")\
                      .country.value_counts().head(5)
max_country_bookings

PRT    21071
GBR     9676
FRA     8481
ESP     6391
DEU     6069
Name: country, dtype: int64

**Вывод:** Топ-5 стран из которых пользователи совершили наибольшее число успешных бронирований:
PRT, GBR, FRA, ESP, DEU

In [8]:
# На сколько ночей (stays_total_nights)  в среднем бронируют отели типа City Hotel? Resort Hotel? 

stays_total_nights_city_hotel= round(bookings.groupby('hotel').agg({'stays_total_nights': 'mean'}), 2)
stays_total_nights_city_hotel

stays_total_nights
hotel                           
City Hotel                  2.98
Resort Hotel                4.32

**Вывод:** Среднее число забронированных ночей в отелях типа city hotel: 2.98. 
В resort hotel: 4.32

Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type).
Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?
*отмена бронирования также считается

In [9]:
bookings.query('assigned_room_type != reserved_room_type').shape[0]


14917

**Вывод:** Всего таких наблюдений 14917

Проанализировать даты запланированного прибытия (arrival_date_year).
На какой месяц чаще всего оформляли бронь в 2016 году? Изменился ли самый популярный месяц в 2017 году?

In [10]:
month_2016 = bookings.query('arrival_date_year == 2016')\
                   .groupby('arrival_date_month')\
                   .agg({'arrival_date_month':'count'}).idxmax()
month_2016

arrival_date_month    October
dtype: object

In [11]:
month_2017 = bookings.query('arrival_date_year == 2017')\
                   .groupby('arrival_date_month')\
                   .agg({'arrival_date_month':'count'}).idxmax()
month_2017 

arrival_date_month    May
dtype: object

****

     Выяснено, что:
     - в 2016 чаще всего бронировали отель на октябрь
     - в 2017 чаще всего бронировали отель на май

****

Сгруппировать данные по годам, а затем проверить, на какой месяц (arrival_date_month)
бронирования отеля типа City Hotel отменялись чаще всего в 2015? 2016? 2017? 

In [12]:
bookings.query('is_canceled == 1 &  hotel == "City Hotel"')\
        .groupby(['arrival_date_year', 'arrival_date_month'])\
        .is_canceled.sum()\
        .groupby(['arrival_date_year']).nlargest(1)

arrival_date_year  arrival_date_year  arrival_date_month
2015               2015               September             1543
2016               2016               October               1947
2017               2017               May                   2217
Name: is_canceled, dtype: int64

****

   **Вывод:** Бронирования отеля типа City Hotel отменялись чаще всего:
    в сентябре 2015 года,
    в октябре 2016 года и
    в мае 2017 года

****

In [13]:
# Какая из трёх колонок: adults, children и babies имеет наибольшее среднее значение?

bookings[['adults', 'children', 'babies']].mean()

adults      1.856403
children    0.103890
babies      0.007949
dtype: float64

**Вывод:** Наибольшее среднее значение у категории adults

Создать колонку total_kids, объединив столбцы children и babies. 
Для отелей какого типа среднее значение переменной оказалось наибольшим?

City hotel – отель находится в городе
Resort hotel – отель курортный
В качестве ответа указать наибольшее среднее total_kids, округлив до 2 знаков после точки.

In [14]:
bookings['total_kids'] = bookings['children'] + bookings['babies']

bookings.groupby(['hotel']).total_kids.mean().round(2)

hotel
City Hotel      0.10
Resort Hotel    0.14
Name: total_kids, dtype: float64

**Вывод:** Для отеля Resort Hotel среднее значение переменной оказалось наибольшим (0.14)

Не все бронирования завершились успешно (is_canceled), поэтому можно посчитать, сколько клиентов было потеряно в процессе, 

т.е. посчитать метрику под названием Churn Rate.

Churn Rate - это процент клиентов, которые отменили бронирование. 

Необходимо создать переменную has_kids, которая принимает значение True, 

если клиент при бронировании указал хотя бы одного ребенка (total_kids), 

в противном случае – False. Далее проверить, среди какой группы пользователей показатель оттока выше. 

В качестве ответа указать наибольший %, округленный до 2 знаков после точки.

In [15]:
bookings['has_kids'] = bookings.total_kids > 0

In [16]:
without_kids = bookings.query('has_kids == False & is_canceled==1').shape[0]/bookings.query('has_kids == False').shape[0]
without_kids = round(without_kids*100, 2)
without_kids

37.22

In [17]:
with_kids = bookings.query('has_kids == True & is_canceled==1').shape[0]/bookings.query('has_kids == True').shape[0]
with_kids = round(with_kids *100, 2)
with_kids

34.92

**Вывод:** Показатель оттока выше у клиентов без детей (37.22)